### PR TITLE
Stabilized webgl_gpgpu_water simulation 

### LINKED ISSUE #14800: webgl_gpgpu_water example / Mesh height increasing
In the example https://threejs.org/examples/webgl_gpgpu_water.html,
the water mesh is slowly going up.

It takes a lot of time (~ 1/2 hr),  and the mouse cursor must be hovering the mesh and animating the surface the whole time (no hovering --> no problem)

to have continuons mouse data input I added a 'mouse simulator' in Iine 553

```
else{
	fakeX+=0.1;
	fakeY+=0.02;
	uniforms.mousePos.value.set( 100*Math.cos(fakeX), 100*Math.cos(fakeY) );
}

```
The problem is, when adding various cursor positions (I needed 5 points touching the water mesh independently), the water mesh is going up quickly (~10 mins).

Any idea about what could I do to fix this ?

### PR DESCRIPTION
**Pull request by yomboprime**
Fixed increasing water height level.

Fixes #14800

### CODE REVIEW

File: examples/webgl_gpgpu_water.html [hunk 1]
```diff
@@ -122,6 +126,67 @@
 			}
 
 		</script>
+		
+		<!-- This is a 'compute shader' to calculate the current volume of water at the center -->
```
Comment (WestLangley) [inline examples/webgl_gpgpu_water.html]: You mean current _level_ of the water.

There are multiple occurrences of the word 'volume' in this PR.

File: examples/webgl_gpgpu_water.html [hunk 2]
```diff
@@ -550,6 +657,30 @@
...
+					numFrames = 0;
+				
+					heightCompensation = - readWaterLevel();
+
+				}
+
+				// Apply gradually height compensation to reset water level
+				if ( heightCompensation !== 0 ) {
+				
+					var decremHeight = Math.min( 0.2, Math.abs( heightCompensation ) ) * Math.sign( heightCompensation );
```
Comment (WestLangley) [inline examples/webgl_gpgpu_water.html]: What is the reason for 0.2 ?

Comment (yomboprime) [reply examples/webgl_gpgpu_water.html]: It is just the fixed amount of drain every frame

File: examples/webgl_gpgpu_water.html (hunk 1 above)
Comment (yomboprime) [reply examples/webgl_gpgpu_water.html]: Fixed.

File: examples/webgl_gpgpu_water.html (hunk 2 above)
Comment (yomboprime) [reply examples/webgl_gpgpu_water.html]: The exact amount is computed (`heightCompensation`), but applied progressively in each frame (`decremHeight`) after the reading so there is no visual bump in the water level.

You can test by editing `var decremHeight = heightCompensation; heightCompensation = 0;` and will see the visual bump each 2 seconds (or longer period if you want)

Comment (WestLangley) [reply examples/webgl_gpgpu_water.html]: As I said, why not decrement by height / 1200 progressively every frame for 1200 frames, then recompute hight and repeat.

Why does your formula contain `Math.min( 0.2, ... )` ?

Comment (yomboprime) [reply examples/webgl_gpgpu_water.html]: Now I understand, it makes sense. I've done a commit.

Comment (WestLangley) [reply examples/webgl_gpgpu_water.html]: I am not going to be able to invest any more time here...

Perhaps you will agree this is better.

```javascript
if ( ++ numFrames > 120 ) {

	numFrames = 0;

	// Apply gradually height compensation to reset water level

	heightmapVariable.material.uniforms.heightCompensation.value = - readWaterLevel() / 120;

}
```

Comment (yomboprime) [reply examples/webgl_gpgpu_water.html]: I'm sorry, but that would apply only one every 120 frames so it wouldn't be gradual. Also, the drain amount would be 1/120 times the needed one.

**Comment by mrdoob** [comment]
I have the feeling that porting the original water effect to glsl may result in simpler code and a more satisfying and stable simulation.

https://web.archive.org/web/20080618181901/http://freespace.virgin.net/hugo.elias/graphics/x_water.htm

I ported it some years ago to Canvas2D:

https://mrdoob.com/lab/javascript/effects/water/02/

@yomboprime would you like to give it a go?

**Comment by yomboprime** [comment]
I've done it and it looks nice so I've deleted the old algorithm. The waves propagates differently (perhaps just faster) I think I could vary the waves speed by multiplying with a factor in the equations.

I've also added some tennis balls with dynamics because I wanted to keep the read back of textures, just because it is instructive for the reader.

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**